prow/github/client.go

```diff
@@ -2855,27 +2855,63 @@ func (c *client) ListTeamMembers(id int, role string) ([]TeamMember, error) {
 	if c.fake {
 		return nil, nil
 	}
-	path := fmt.Sprintf("/teams/%d/members", id)
+
+	// the "node_id" field is the same as the v4 "id" field, which encodes
+	// a numeric team ID as base64("04:Team%d"). Reverse this for now, until
+	// We can convert all the IDs to node_ids.
+	teamNumberAsId := base64.StdEncoding.EncodeToString([]byte(fmt.Sprintf("04:Team%d", id)))
+
+	var q struct {
+		Node struct {
+			ID   string `graphql:"id"`
+			Team struct {
+				Members struct {
+					Edges []struct {
+						Node struct {
+							Login string
+						}
+						Role string
+					}
+					PageInfo struct {
+						EndCursor   *githubql.String
+						HasNextPage bool
+					}
+					TotalCount int
+				} `graphql:"members(first: 100, after: $memberCursor, membership: IMMEDIATE)"`
+			} `graphql:"... on Team"`
+		} `graphql:"node(id:$id)"`
+	}
+	vars := map[string]interface{}{
+		"id":           githubql.ID(teamNumberAsId),
+		"memberCursor": (*githubql.String)(nil), // Nil to get first page
+	}
+	// The v4 API uses uppercase constants for MAINTAINER and MEMBER, and does not support
+	// the ALL role. To work around this, we filter by role explicitly in the code below so
+	// that we don't need to create two different queries.
+	// Ideally at some point, we could add a role to the TeamMember struct and reduce our API
+	// calls...
+	role = strings.ToUpper(role)
+
 	var teamMembers []TeamMember
-	err := c.readPaginatedResultsWithValues(
-		path,
-		url.Values{
-			"per_page": []string{"100"},
-			"role":     []string{role},
-		},
-		// This accept header enables the nested teams preview.
-		// https://developer.github.com/changes/2017-08-30-preview-nested-teams/
-		"application/vnd.github.hellcat-preview+json",
-		func() interface{} {
-			return &[]TeamMember{}
-		},
-		func(obj interface{}) {
-			teamMembers = append(teamMembers, *(obj.(*[]TeamMember))...)
-		},
-	)
-	if err != nil {
-		return nil, err
+	for {
```

Comment:
@cjwagner how does ghproxy work with graphql?

Reply:
It doesn't cache, graphql requests are POSTed queries rather than simple REST requests and they don't support ETag based caching. Using graphql is only more efficient than v3+ghproxy in cases where the number of requests is small, there are a large number of resources, and the resources change frequently. Since the number of requests for this endpoint is relatively large and teams don't change frequently I suspect the v3 API + ghproxy will use less of the API ratelimit.

Reply:
What about collecting all the teams and members in a single/paged set of queries? For peribolos, this would probably collapse N*(team, team members * 2) to N*(members per team/100).

Reply:
The resource we try to conserve here is uncached requests. So right now peribolos is essentially free because essentially 100% of the requests are cached.

AKA 10,000 cached requests is cheaper to continually run (cost ~0) than 10 uncached requests (cost 10 per run).

So ideally we keep the cost at C, not N*C which this PR does right now.

Looking through the docs though it looks like github makes it intentionally hard/impossible to only get direct members https://developer.github.com/v3/teams/members/

Could we make this a flag? AKA by default we continue using the v3 api and including children members in the parent with a flag to make this more precise by using more tokens and the v4 api?

Reply:
It looks like GitHub doesn't provide an equivalent GET endpoint for GraphQL, despite the standard recommending it. You *could* enable caching for POST queries to the graphql endpoint if they have the "query" key in the JSON body, but I agree that seems like a lot more work.

I don't know if it matters that the V4 and V3 APIs have [separate rate limits](https://developer.github.com/v4/guides/resource-limitations/#:~:text=The%20GraphQL%20API%20v4%20limit,per%20hour%20for%20authenticated%20requests).

I'm moving this to a separate "ListTeamsV4" function with a new return type -- I agree that it's annoying that you can't get this information from the V3 API.